src/client.ts

```diff
@@ -76,7 +78,128 @@ export default function connect(
   dsn?: string | ConnectConfig | null,
   options?: ConnectConfig | null
 ): Promise<Connection> {
-  return ConnectionImpl.connect(dsn, options);
+  let config: ConnectConfig | null = null;
+  if (typeof dsn === "string") {
+    config = {...options, dsn};
+  } else {
+    if (dsn != null) {
+      // tslint:disable-next-line: no-console
+      console.warn(
+        "`options` as the first argument to `edgedb.connect` is " +
+          "deprecated, use " +
+          "`edgedb.connect('instance_name_or_dsn', options)`"
+      );
+    }
+    config = {...dsn, ...options};
+  }
+  return StandaloneConnection.connect(parseConnectArguments(config));
+}
+
+function sleep(durationMicros: number): Promise<void> {
+  return new Promise((accept, reject) => {
+    setTimeout(() => accept(), durationMicros);
+  });
+}
+
+class StandaloneConnection implements Connection {
+  [ALLOW_MODIFICATIONS]: never;
+  private config: NormalizedConnectConfig;
+  private _connection?: ConnectionImpl;
+  private _isClosed: boolean; // For compatibility
+
+  private constructor(config: NormalizedConnectConfig) {
+    this.config = config;
+    this._isClosed = false;
+  }
+  async transaction<T>(
+    action: () => Promise<T>,
+    options?: TransactionOptions
+  ): Promise<T> {
+    return await (await this.connection()).transaction(action, options);
+  }
+  async close(): Promise<void> {
+    try {
+      if (this._connection) {
+        await this._connection.close();
+      }
+      // TODO(tailhook) it makes little sense to close the reconnecting
+      // connection so maybe deprecate this method
+      this._isClosed = true;
+    } finally {
+      this._cleanupProxy();
+    }
+  }
+  private _cleanupProxy(): void {
+    const proxy = proxyMap.get(this);
+    if (proxy != null) {
+      proxy[onConnectionClose]();
+      proxyMap.delete(this);
+    }
+  }
+  isClosed(): boolean {
+    return this._isClosed;
+  }
+  async execute(query: string): Promise<void> {
+    return await (await this.connection()).execute(query);
+  }
+  async query(query: string, args?: QueryArgs): Promise<Set> {
+    return await (await this.connection()).query(query, args);
+  }
+  async queryJSON(query: string, args?: QueryArgs): Promise<string> {
+    return await (await this.connection()).queryJSON(query, args);
+  }
+  async queryOne(query: string, args?: QueryArgs): Promise<any> {
+    return await (await this.connection()).queryOne(query, args);
+  }
+  async queryOneJSON(query: string, args?: QueryArgs): Promise<string> {
+    return await (await this.connection()).queryOneJSON(query, args);
+  }
+  async connection(): Promise<ConnectionImpl> {
+    if (this._isClosed) {
+      throw new errors.InterfaceError("Connection is closed");
+    }
+    if (this._connection && !this._connection.isClosed()) {
+      return this._connection;
+    }
+    const max_time = process.hrtime.bigint()
+      + BigInt(Math.ceil((this.config.waitUntilAvailable || 0) * 1000_000));
```

Comment:
Nit, imo this reads better

```suggestion
      + BigInt(Math.ceil((this.config.waitUntilAvailable || 0) * 1_000_000));
```

Reply:
But why do you want to use high-res bigint timing here? IMO the precision of milliseconds is more than enough here, no?

Reply:
It's the only monotonic clock available as far as I know.

Reply:
Ah, makes sense. Could you please add a comment explaining that and mentioning that hrtime is monotonic?